### ISSUE TITLE
Optimizing control flow?

### ISSUE THREAD

**Issue by kripken** [pipcet/asmjs#7]
I think that this project emits a switch-in-a-loop pattern for control flow currently? If that's correct, then one option to make it fast could be to add a pass in the binaryen optimizer to handle that (it does have the "relooper" implementation, that goes from cfgs to loops/ifs, but it doesn't parse switches in loops into a cfg which would be necessary here). If that's potentially useful I could look into adding that pass in binaryen.

**Comment by pipcet** [comment]
> I think that this project emits a switch-in-a-loop pattern for control flow currently?

Yes, that's right. We definitely want to make use of the relooper algorithm at some point to improve performance; I don't know my way around binaryen well enough yet, so any thoughts or help would be much appreciated!

> If that's potentially useful I could look into adding that pass in binaryen.

I have no clue how difficult it would be, but if it's workable that would be extremely useful to me!

Thanks for taking the time to write about this, do feel free to share any other thoughts or ideas you might have!

**Comment by kripken** [comment]
Ok, so one option would be to integrate the relooper in your compiler. That would be more efficient overall, but take more work on your side.

The other is as suggested here, to add support in binaryen for the loop-switch pattern. In fact we should add general support for recreating CFGs, but the loop-switch is common enough that we should support it specifically - and in particular detect the special variable used to thread control flow. That does mean matching the more specific pattern of setting the variable and so forth, so we'd need to coordinate that probably. But I think it's worth doing, and not too hard. What specifically is the pattern you emit?

**Comment by pipcet** [comment]
> What specifically is the pattern you emit?

It's probably as simple as possible: there's a fixed local (it's actually local 0) that we're feeding into `br_table`, and for branches we set it to an `i32.const` and `br` to the loop, either directly or in an `if`:

```
  (block $label$0
   (loop $label$1
    (block $label$2
     (block $label$3
      (block $label$4
       (block $label$5
        (br_table $label$2 $label$4 $label$3 $label$5
         (get_local $var$0)
        )
```

```
        (block $label$6
         (set_local $var$0
          (i32.const 2)
         )
         (br $label$1)
        )
```

**Comment by kripken** [comment]
Thanks, that's basically what I'd expect. I opened https://github.com/WebAssembly/binaryen/issues/953 for this.

**Comment by pipcet** [comment]
I'm playing with the relooper code (can't say I understand it yet :-), so it'll be a while before it really can be of any help). Am I right in thinking that it treats wasm code essentially as opaque? I'm wondering how hard it would be to dump "pre-assembly" code that contains (as far as the relooper is concerned) blocks of text split into blocks and branches between such blocks, and run the relooper on that...

**Comment by ghost** [comment]
@pipcet There are much simpler algorithms around that do much of the work that the relooper does in a page or two of code, so organising the CFG into islands of loops and nesting them, perhaps a google will find the.

From what I understand the relooper will not help unless the CFG loops are reducible, and that it does not eliminate irreducible loops. I agree that it would be better to have a clear pattern, perhaps a separate loop operator, for the loop-switch pattern that makes the dispatch variable clear to save the consumer having to match this pattern. 

I see that binaryen now seems to see some merit in an SSA encoding which sounds encouraging, and might you even consider passing SSA encoder code to binaryen to be packed and encoded? While converting the wasm blocks to SSA is trivial, I have not found it so trivial for loops. For loops it seems to take some effort to eliminate unnecessary phi nodes. For example look at the V8 decoder, it assumes all local variables are phi nodes to loops, with an extra pass making some attempt to reduce this but it's still not minimal. So if you could emit the loops in SSA with a minimal set of phi nodes it might really help consumers such as binaryen.

I noticed that the llvm backend generated many irreducible loops where the emscripten pipeline did not. One guess is that it is due to jump threading optimizations, jumping into a loop. Some gcc optimizations that work for native code might not be good for the wasm style control flow, and is that an area you could take a look at.

The loop-switch pattern does not appear to work very well with SSA. I see a lot of phi nodes on the loops with many not changing on a particular path. This is one area that the local variable help as only a few are typically modified on a particular path around this loop-switch, but it would appear better if there were block scoped mutable variables.

Might you also consider keeping some meta information about loop induction variables in the encoding. Does gcc keep this information around to use? It is a lot of work for the consumer to analyse the loops again. Just having the stride might be a big help as it enables a range of optimizations to eliminate bounds checks even if the limits of the induction variables are dynamic.

One thought is to define a loop variable that counts from zero and increases by one on each loop, and then to define other strides in expressions based on this. The consumer would be expected to be smart enough to derive other strides from this, but that appears a much simpler task. So add a 'loop_with_stride' operator to make this explicit. There seem to be many other options here.

I suggest also avoid that nested-block-br_table, and a br_case operator seems to be cleaner and I think that will be clearer as the SSA encoding is explored as br_table passes a fixed set of definitions to the successors but in SSA these successors will be accepting different sets of definitions in general. This problem is already visible if one of the br_table targets is the function return with not  value and another wants a value.

**Comment by kripken** [comment]
@pipcet: yes, the relooper impl in binaryen treats the wasm contents of blocks as opaque. And yeah, dumping some kind of "pre-assembly" is a good idea I think, I've called that "subwasm" when discussing it, basically it's wasm without structured control flow. If we standardized that, then tools like gcc and clang could emit subwasm easily, and tools like binaryen would structure and optimize the output into wasm.

@wllang it does take more than a page or two of code to get good results, though (the wasm backend uses such a method, and it's one reason it's output isn't as good as it could be).

**Comment by ghost** [comment]
Here's one algorithm that claims good results and to be fast http://lenx.100871.net/papers/loop-SAS.pdf and a page or two of code. At least this might help understand the type of algorithms available, and be another perspective on what the relooper is doing.

Wasm can encode DAG CFGs naturally and perhaps the relooper had different aims that added complexity. Some extra code is needed to actually emit the irreducible loops, and there are still often choices to be made in DAG ordering.

It seems a hard problem for consumers such as binaryen to undo jump threading and other code transforms that generate irreducible loops, so perhaps it would help the producer understand the challenge to at least preserve the loop structure in the IR, and as noted it will save binaryen some work if the induction variables are communicated in the IR, and to have the minimal set of loop phis, and for the loop-switch pattern to be explicit.

Even if binaryen does add loop analysis it will add to the compilation time, so if the producer can emit an IR with this already done than that might speed compilation.

**Comment by kripken** [comment]
@wllang: that paper recreates loops, but you also need to recreate ifs and if-elses for good code size (that's the problem with llvm output now - it has loops, but not if-elses). The relooper does precisely that - recreate loops and if-elses, nothing extra. The core relooper algorithm is pretty small (1-2 pages in [the paper](https://raw.githubusercontent.com/kripken/emscripten/master/docs/paper.pdf)); the size of the current relooper code is because it has a bunch of optimizations to get good results.

I agree that binaryen doing loop/if analysis will add to compile time. Still I think it's a useful pass to add to binaryen - it will be useful not just for current gcc and llvm, but other compilers too, and may also help in later stages of compilation anyhow (after some optimization passes, the cfg may be further optimized).

**Comment by ghost** [comment]
Looking at this further suggests it will not be possible to convert this to static CFG because the switch value is a dynamic function argument, see https://github.com/pipcet/asmjs/blob/everything/asmjs.org

It's an interesting design but I can't see how it could perform well. It appears that the design is trying to stretch to support setjmp and longjmp but at a high cost of having to spill values to the linear memory and to pass function return values in the linear memory. It would appear more productive to just work on getting some support to unwind the stack into wasm, and then to work with the wasm design rather than working around it.

The design also appears to use a limited number of local variables and spills to the linear memory. It would be better to try to use local variables or the stack to the full, so only place values in the linear memory if their address is taken.

**Comment by pipcet** [comment]
@wllang Indeed, that design is something I'm unhappy with; it's by no means set in stone. Two minor points:

It's not just setjmp/longjmp: the reason for the unusual design is the total lack of asynchronous function calling capabilities in the MVP.

I do disagree that it would necessarily be better to use local variables instead of stack slots in all circumstances, though; there is a limited number of physical registers, and spilling to linear memory isn't that much more expensive than spilling to the VM stack. I think it's possible for the better register allocation that gcc does to shift the balance in favor of a limited number of local variables which are then allocated with essentially no register pressure.

Again, though, both of those are up for changing.

**Comment by ghost** [comment]
@pipcet If definitions spill to the linear memory then any memory access could modify the value and this is not good for the compiler - it means that their derived type becomes invalid on any linear memory access or call, and that hits areas like alignment information and bounds check elimination and loop analysis.

I would go even further and claim that if a definition is written to linear memory and reused then is should be kept live in the local variable or values stack rather than being re-loaded from linear memory when needed. For example a stack pointer to a linear memory stack should be kept in a local variable or the values stack, writing back to linear memory or a global where it might be needed by a callee, but always using the value in the local variable in uses within the function. This allows a range check of the stack pointer to be hoisted to the start of the function.

The consumer can unroll definitions in local variables into SSA form, but not definitions in linear memory (not without making some assumptions about the generated code style, and even then it becomes really hard work if the address of an element in the linear memory stack is taken).

The idea of doing register allocation in the producer seems a dead end to me, and I think wasm would have been better just removing local variables. All the values on the wasm stack need to be allocated somewhere too, so the live range of the locals and the stack values both need to be computed anyway and both allocated to registers. Integrating the wasm decoder with a register allocator seems a key area or future study, and when done well I think we will see some really fast wasm compilers generating great code - and any values spilled to the linear memory will be dead to optimization by these compilers.
